### ISSUE TITLE
Investigate using `BigInt`'s with `< 4` limbs for `FieldElement`'s

### ISSUE THREAD

**Issue by michaeljklein** [noir-lang/noir#6316]
### Problem

`bn254` `FieldElement`'s are defined in [`acir_field`](https://github.com/noir-lang/noir/blob/4274efada969847e338e56793701c2f988c6f42c/acvm-repo/acir_field/src/lib.rs#L18)

```rust
pub type FieldElement = field_element::FieldElement<ark_bn254::Fr>;
```

Where [`field_element::FieldElement`](https://github.com/noir-lang/noir/blob/4274efada969847e338e56793701c2f988c6f42c/acvm-repo/acir_field/src/field_element.rs#L13) wraps its argument and ensures it's a `PrimeField`:

```rust
pub struct FieldElement<F: PrimeField>(F);
```

[`ark_bn254::Fr`](https://docs.rs/ark-bn254/latest/ark_bn254/type.Fr.html) is defined as:

```rust
pub type Fr = Fp256<MontBackend<FrConfig, 4>>;
```

Where [`Fp256`](https://docs.rs/ark-ff/latest/ark_ff/fields/models/fp/type.Fp256.html) defines the contents (`MontBackend..` defines the `FpConfig`):

```rust
pub type Fp256<P> = Fp<P, 4>;
```

And [`Fp`](https://docs.rs/ark-ff/latest/ark_ff/fields/models/fp/struct.Fp.html):

```rust
pub struct Fp<P: FpConfig<N>, const N: usize>(pub BigInt<N>, _);
```

Is an [`N`-limbed `BigInt`](https://docs.rs/ark-ff/latest/ark_ff/biginteger/struct.BigInt.html):

```rust
pub struct BigInt<const N: usize>(pub [u64; N]);
```

So our `bn254` `FieldElement`'s are always represented by `[u64; 4]`, even when their values are much smaller than `u64::MAX`

### Happy Case

Investigate client-size memory savings and performance overhead of using a flexibly-sized big integer type instead of `BigInt<N>`, e.g. a [BigInt](https://crates.io/crates/num-bigint) that uses a `Vec` of limbs

### Workaround

None

### Project Impact

Nice-to-have

### Would you like to submit a PR for this Issue?

None

**Comment by TomAFrench** [comment]
Isn't this expected however? We have 4 u64s because it's a bn254 field representation, if you provide a different `AcirField` implementation for another field modulus then it'll be backed by a different number of limbs, etc.

**Comment by michaeljklein** [comment]
@TomAFrench yes, this is somewhat expected because we can need `4` `u64`'s to represent a `bn254` element, but we don't need exactly `4` `u64`'s for smaller field elements. For example, [GMP](https://gmplib.org/manual/Integer-Internals) uses a variable number of limbs that are dynamically (re)allocated as needed. E.g. `8` uses only one limb, `2^65` uses two limbs, etc.

**Comment by TomAFrench** [comment]
My point is that you can use anything which satisfies the `AcirField` trait and we're not enshrining a particular number of limbs (until you get further up into nargo which isn't yet generic)
